src/components/legend.ts

```diff
@@ -237,88 +514,107 @@ namespace Plottable.Components {
+          const withinColumn = within(p, columnBounds);
+
+          if (withinColumn) {
+            const rowElement = this.content().selectAll(`.${Legend.LEGEND_ROW_CLASS}`)[0][rowIndex];
+            // HACKHACK The 2.x API chooses the symbol element as the "selection" to return, regardless of what
+            // was actually selected
+            const entryElement = d3.select(rowElement)
+              .selectAll(`.${Legend.LEGEND_ENTRY_CLASS}`)[0][Math.floor(columnIndex / 2)];
+             const symbolElement = d3.select(entryElement).select(`.${Legend.LEGEND_SYMBOL_CLASS}`);
+
+            // HACKHACK The 2.x API returns the center {x, y} of the symbol as the position.
+            const rowTranslate = d3.transform(d3.select(rowElement).attr("transform")).translate;
+            const symbolTranslate = d3.transform(symbolElement.attr("transform")).translate;
+
+            return [{
+              datum: column.getData(),
+              position: { x: rowTranslate[0] + symbolTranslate[0], y: rowTranslate[1] + symbolTranslate[1] },
+              selection: d3.select(entryElement),
+              component: this,
+            }];
           }
-          lowX += layout.entryLengths.get(value);
-        });
-      });
 
-      return entities;
+          return entity;
+        }, entity);
+
+      }, [] as Entity<Legend>[])
     }
 
     public renderImmediately() {
       super.renderImmediately();
-
-      let layout = this._calculateLayoutInfo(this.width(), this.height());
-
-      let rowsToDraw = layout.rows.slice(0, layout.numRowsToDraw);
-      let rows = this.content().selectAll("g." + Legend.LEGEND_ROW_CLASS).data(rowsToDraw);
-      rows.enter().append("g").classed(Legend.LEGEND_ROW_CLASS, true);
-      rows.exit().remove();
-
-      rows.attr("transform", (d: any, i: number) => "translate(0, " + (i * layout.textHeight + this._padding) + ")");
-
-      let entries = rows.selectAll("g." + Legend.LEGEND_ENTRY_CLASS).data((d) => d);
-      let entriesEnter = entries.enter().append("g").classed(Legend.LEGEND_ENTRY_CLASS, true);
-      entriesEnter.append("path");
-      entriesEnter.append("g").classed("text-container", true);
-      entries.exit().remove();
-
-      let legendPadding = this._padding;
-      rows.each(function (values: string[]) {
-        let xShift = legendPadding;
-        let entriesInRow = d3.select(this).selectAll("g." + Legend.LEGEND_ENTRY_CLASS);
-        entriesInRow.attr("transform", (value: string, i: number) => {
-          let translateString = "translate(" + xShift + ", 0)";
-          xShift += layout.entryLengths.get(value);
-          return translateString;
+      const table = this._buildLegendTable(this.width(), this.height());
+      const entryNames = this._colorScale.domain().slice().sort((a, b) => this._comparator(this._formatter(a), this._formatter(b)));
+
+      // clear content from previous renders
+      this.content().selectAll("*").remove();
```

Comment:
This isn't really the idiomatic way to do d3. Removing and re-adding everything on every render causes lots of DOM thrashing. You can see from the code that this replaces that it applies the rows to the DOM using the `.data` on the selection then using `.enter`, `.exit` and the selection itself to update the stuff that's there.